### ISSUE TITLE
Define initial pledge computation for genesis state

### ISSUE THREAD

**Issue by anorth** [filecoin-project/specs-actors#707]
The genesis state will have some already-sealed power. Since the pledge computation is a function of the existing sealed power, it's not well defined for the all-at-once genesis power.

@zixuanzh @davidad we need your help to define something sensible here.

**Comment by davidad** [comment]
The obvious proposal would be to just exempt genesis power from an initial pledge requirement, i.e., set the stored IP in the genesis state for pre-sealed sectors to 0. Any reason that doesn’t work?

**Comment by magik6k** [comment]
Lotus genesis creation mostly reuses miner actor Precommit/Commit flow to get sectors into the state, which means that most solutions would require special-casing things in actors

See https://github.com/filecoin-project/lotus/blob/next/chain/gen/genesis/miners.go#L242-L257

**Comment by davidad** [comment]
Got it. So I suppose that InitialPledgeForPower should return 0 at genesis—but __only__ at genesis, not just if total power happens to be zero (which could hypothetically happen later). I’d kind of like to have access to chainHeight in InitialPledgeForPower anyway, which makes me want to suggest that the special case condition should be chainHeight==0 (genesis by definition), but maybe there is some other condition that is both necessary and sufficient?

**Comment by anorth** [comment]
Quoth @zx:
> Total initial pledge at genesis should be computed from the same initial pledge calculation with genesis state, using initial values for alpha-beta filters. It is then divided pro rata based on Quality Adjusted power for each sector. Initial pledge for each genesis miner will be the sum of IPs for all their sectors.

**Comment by anorth** [comment]
The problem here is in the initial issue description: pledge computation is a function of the existing sealed power, which is not well defined during the calculation of genesis state.

But what I think this is requesting is to pretend that all the genesis power was already committed, then calculated initial pledge for each sector. Or to compute the per-sector pledge as if this genesis power were being committed as an atomic unit (which is not possible in normal operations).

We might need some funky special case methods in the actors to do this, but if this is the stable spec @zixuanzh, we can get on it.

**Comment by zixuanzh** [comment]
I will pretend that all genesis power has already been committed and then calculate initial pledge for each sector. However, we have yet to fully specify network initialization. This is our current best guess.

**Comment by anorth** [comment]
I think this is as un-blocked as we're going to get it, for now; ready for implementation.

**Comment by ZenGround0** [comment]
Since this issue was filed we now create reward and power actor state with initial filter values that estimate the total genesis power, reward and rates of change of power and reward.  This approach appears to already have solved the problem.

The sector commitments during lotus genesis construction are done from [SetupStorageMiners](https://github.com/filecoin-project/lotus/blob/next/chain/gen/genesis/genesis.go#L408) which is called only after setting up [reward](https://github.com/filecoin-project/lotus/blob/next/chain/gen/genesis/genesis.go#L147) and [power](https://github.com/filecoin-project/lotus/blob/next/chain/gen/genesis/genesis.go#L167) states.

The power and reward filters are then taken directly from the already setup actors (using [currentTotalPower](https://github.com/filecoin-project/lotus/blob/next/chain/gen/genesis/miners.go#L327) and [currentEpochBlockReward](https://github.com/filecoin-project/lotus/blob/next/chain/gen/genesis/miners.go#L365) ).  

This means that @zixuanzh's spec:
> Total initial pledge at genesis should be computed from the same initial pledge calculation with genesis state, using initial values for alpha-beta filters. It is then divided pro rata based on Quality Adjusted power for each sector.

is the current behavior.

One last thing -- right now the initial genesis power is an estimate hard coded as a constant in the power actor code in specs actors.  If we want the genesis IP calculation to be precise we will either need to tweak this constant to match genesis power before every network start or come up with a more configurable approach.

**Comment by anorth** [comment]
This sounds good, but I have not dived in to all the related code to confirm for myself.

It sounds like there's a circular dependency which I'm not clear how is resolved. You say that the reward and power states are set up before the miners,  but also that the power and reward filters are taken directly from the "already set up actors". But the miners are not set up in the sense that they don't have pledged sectors yet?

Is the total amount of storage that _will be_ pledged in the genesis state calculated somewhere up front, in order to initialise the values that will be inputs into the pledge calculation when those sectors _are_ added? I might be just misunderstanding something.

**Comment by ZenGround0** [comment]
> You say that the reward and power states are set up before the miners, but also that the power and reward filters are taken directly from the "already set up actors".

The filter power value is just a guess we hard code in the actor code.  There is nothing that sets it to the exact value of the genesis power. 

> Is the total amount of storage that will be pledged in the genesis state calculated somewhere up front, in order to initialize the values that will be inputs into the pledge calculation when those sectors are added?

It *could* be, right now we always assume 750 TB.  I kept this open to figure out if we want to do this exactly.  One easy way is to set the parameter in a PR for every reset.  Another is to set the specs actors parameter from genesis setup.  If we want something else we'll need to do some design.

**Comment by anorth** [comment]
Ok, I understand now, and think this we're good. To clarify

```
InitialPledgeForPower(power, 
	rewardStats.ThisEpochBaselinePower,  // fixed function
	pwrTotal.PledgeCollateral, // unused in the computation
	rewardStats.ThisEpochRewardSmoothed, // filter value
	pwrTotal.QualityAdjPowerSmoothed,  // filter value
	circulatingSupply  // provided by runtime
)
```

CE have previously requested that we retain the network pledge collateral total as a parameter to this function, but it's still unused. I think it's role is now taken care of by the circulating supply value, which already subtracts this. @zixuanzh ?

This circulating supply value must be provided by the runtime, and I'm expecting the Lotus genesis runtime to return `0` for this value during genesis (at least, for networks like mainnet with no faucet). @arajasek

**Comment by ZenGround0** [comment]
Reopening because ongoing discussion suggests we might want to do something beyond hardcoding either here or in genesis construction.

**Comment by davidad** [comment]
I’m (weakly) in favor of:
1. In an early phase of genesis construction, compute in advance the amount of QA power that *will* be added. (I am not sure if this is currently done or needs more implementation.)
2. Set the initial smoothed position estimate for total network QA power to this value, instead of a hard-coded 750TiB as it is currently.
3. Then compute the pledge using that value (no change from how it’s done now, if I understand correctly).

This has two main advantages:
1. In case there are late-breaking changes to the amount of genesis power for mainnet, it will be nice to not have to worry about modifying a hard-coded value in specs-actors and propagating that release before launch.
2. When setting up small devnets, the cryptoecon values that depend on total network power will be more reasonable.

**Comment by anorth** [comment]
No longer relevant.
